lib/index.js

```diff
@@ -1,44 +1,124 @@
+  app.get('/heartbeat', Routes.uptime)
+  app.use(Middleware.stats(appConfig.get('statsd')))
+
+  config.routes.forEach((route) => {
+    var url = URL.resolve(route.dhisUrl, route.dhisPath)
+
+    // Populate cache on startup
+    Utils.populateCache(url)
+    Utils.addCronJob(route.cronPattern, route.cronTimezone, url)
+    Utils.addRoute(app, route.dhisPath)
+  })
+  return app
+}
+exports.setupApp = setupApp
+
+// Load the config
+function start (callback) {
+  setupConfig((err, appConfig) => {
+    if (err) {
+      callback(err)
+    }
+
+    let apiConf = appConfig.get('openhim')
+    let port = appConfig.get('server:port')
+
+    if (apiConf.api.trustSelfSigned) {
+      process.env.NODE_TLS_REJECT_UNAUTHORIZED = '0'
+    }
+
+    if (apiConf.register) {
+      medUtils.registerMediator(apiConf.api, mediatorConfig, (err) => {
+        if (err) {
+          Winston.error('Failed to register this mediator, check your config')
+          Winston.error(err.stack)
+          process.exit(1)
+        }
+        apiConf.api.urn = mediatorConfig.urn
+        medUtils.fetchConfig(apiConf.api, (err, newConfig) => {
+          Winston.info('Received initial config:')
+          Winston.info(JSON.stringify(newConfig))
+          let config = newConfig
+          if (err) {
+            Winston.error('Failed to fetch initial config')
+            Winston.error(err.stack)
+            process.exit(1)
+          } else {
+            Winston.info('Successfully registered mediator!')
+            let app = setupApp(config, appConfig)
+            server = app.listen(port, () => {
+              if (apiConf.heartbeat) {
+                let configEmitter = medUtils.activateHeartbeat(apiConf.api)
+                configEmitter.on('config', (newConfig) => {
+                  Winston.info('Received updated config:')
+                  Winston.info(JSON.stringify(newConfig))
+                  // set new config for mediator
+
+                  Utils.updateConfig(app, config.routes, newConfig.routes)
+                })
+              }
+              callback(null, port, server)
+            })
+          }
+        })
+      })
+    } else {
+      // default config
+      let app = setupApp(mediatorConfig.config, appConfig)
+      server = app.listen(port, () => callback(null, port, server))
+    }
+  })
+}
+exports.start = start
+
+function setupConfig (callback) {
+  Confit(Path.join(__dirname, '..', 'config')).create(function (err, config) {
+    if (err) {
+      return callback(err)
+    }
+    callback(null, config)
+  })
+}
+exports.setupConfig = setupConfig
 
-  // Start the server
-  app.listen(config.get('port'), function() {
-    Winston.info('Server started');
-  });
-});
+function stop (callback) {
+  server.on('close', () => {
+    Cache.close(callback)
+  })
+  server.close()
+}
+exports.stop = stop
 
-module.exports = app;
+if (!module.parent) {
+  // if this script is run directly, start the server
+  start((err, port) => {
```

Comment:
Instead of passing the port around you can just use `server.address()` to get it.